File: src/main/erlang/g0301_0400/s0322_coin_change/Solution.erl
% #Medium #Top_100_Liked_Questions #Array #Dynamic_Programming #Breadth_First_Search
% #Algorithm_II_Day_18_Dynamic_Programming #Dynamic_Programming_I_Day_20
% #Level_2_Day_12_Dynamic_Programming #Big_O_Time_O(m*n)_Space_O(amount)
% #2025_01_20_Time_797_(100.00%)_Space_66.72_(100.00%)

-spec coin_change(Coins :: [integer()], Amount :: integer()) -> integer().
coin_change(Coins, Amount) ->
    % Sort coins in ascending order
    SortedCoins = lists:sort(Coins),
    
    % Initialize DP map with base case
    InitDP = #{0 => 0},
    
    % Build DP table for amounts from 1 to target
    case Amount of
        0 -> 0;
        _ ->
            FinalDP = build_dp_table(SortedCoins, lists:seq(1, Amount), InitDP),
            % Get result or return -1 if not possible
            maps:get(Amount, FinalDP, -1)
    end.

% Build DP table for all amounts
-spec build_dp_table([integer()], [integer()], map()) -> map().
build_dp_table(_Coins, [], DP) -> DP;
build_dp_table(Coins, [Amount|Rest], DP) ->
    NewDP = process_coins_for_amount(Coins, Amount, DP),
    build_dp_table(Coins, Rest, NewDP).

% Process all coins for current amount
-spec process_coins_for_amount([integer()], integer(), map()) -> map().
process_coins_for_amount([], _Amount, DP) -> DP;
process_coins_for_amount([Coin|Rest], Amount, DP) ->
    case Coin > Amount of
        true ->
            % Stop if coin is larger than current amount
            DP;
        false ->
            % Try using current coin
            case maps:get(Amount - Coin, DP, undefined) of
                undefined ->
                    % Can't use this coin, try next
                    process_coins_for_amount(Rest, Amount, DP);
                PrevCount ->
                    % Can use this coin, update minimum coins needed
                    NewCount = PrevCount + 1,
                    NewDP = maps:update_with(Amount, 
                                           fun(OldVal) -> min(OldVal, NewCount) end,
                                           NewCount, 
                                           DP),
                    process_coins_for_amount(Rest, Amount, NewDP)
            end
    end.

File: src/main/c/g0301_0400/s0322_coin_change/Solution.c
// #Medium #Top_100_Liked_Questions #Array #Dynamic_Programming #Breadth_First_Search
// #Algorithm_II_Day_18_Dynamic_Programming #Dynamic_Programming_I_Day_20
// #Level_2_Day_12_Dynamic_Programming #Top_Interview_150_1D_DP #Big_O_Time_O(m*n)_Space_O(amount)
// #2024_11_07_Time_22_ms_(87.50%)_Space_12.3_MB_(23.00%)

#include <stdio.h>
#include <limits.h>

int coinChange(int* coins, int coinsSize, int amount) {
    int* dp = (int*)malloc((amount + 1) * sizeof(int));

    // Initialize dp array, where dp[i] represents the minimum number of coins to make amount i.
    // Set dp[0] to 1 as the base case (1 represents that 0 coins are needed to make amount 0).
    dp[0] = 1;
    for (int i = 1; i <= amount; i++) {
        dp[i] = 0;
    }

    for (int c = 0; c < coinsSize; c++) {
        int coin = coins[c];
        for (int i = coin; i <= amount; i++) {
            int prev = dp[i - coin];
            if (prev > 0) {  // If it's possible to make amount (i - coin)
                if (dp[i] == 0) {
                    dp[i] = prev + 1;
                } else {
                    dp[i] = (dp[i] < prev + 1) ? dp[i] : prev + 1;
                }
            }
        }
    }

    int result = (dp[amount] > 0) ? dp[amount] - 1 : -1; // Subtract 1 to account for the base case offset

    free(dp); // Free allocated memory
    return result;
}
